src/shape.cpp

```diff
@@ -151,6 +152,24 @@ struct shape_impl
             m_lens.begin(), m_lens.end(), std::size_t{1}, std::multiplies<std::size_t>());
     }
 
+    std::size_t get_index(size_t i) const
```

Comment:
Why is this moved to the impl? The impl class doesnt use this.

Reply:
Because the implementation is heavily involved in all calculations: We are accessing impl, a shared pointer, for every single operand in that calculation of index (which gets repeated millions of times for a large tensor..) 

BTW, just an aside question, why is this impl a shared pointer?  Thanks.

Reply:
> BTW, just an aside question, why is this impl a shared pointer? 

Because its cheaper to ref count then allocate and copy.

Reply:
The shared_pointer implementation of shape.hpp is overall slower by 7.5% than not using it. 

(I compared the driver read operation on bevdet.onnx; In debug mode, as before. The only difference is the workstation is different this time for the apples to apples comparison.)

Reply:
> The shared_pointer implementation of shape.hpp is overall slower by 7.5% than not using it.

So what did you use instead of a `shared_ptr`?

Reply:
> In debug mode

Of course, the debug mode is slow. Its faster to pass `-01` or `-02` to the debug build to get better performance.

Reply:
(All the testing was done in debug mode, so just wanted to mention the comparison numbers. That also likely uses all the asserts to execute (likely turned off in release mode.)

Instead of a shared_ptr, the structure shape_impl was just included in struct shape.. No pointers. I can check that code in after running all the unit-tests.

Reply:
> I can check that code in after running all the unit-tests.

You could put it into a separate branch and we could try to test it, but I would be suprised that it would be faster in release mode.

Reply:
I will create a separate branch for it.. thanks.